x-pack/plugin/esql/src/main/antlr/EsqlBaseParser.g4

```diff
@@ -180,6 +185,18 @@ limitCommand
     : LIMIT INTEGER_LITERAL
     ;
 
+limitCommand_PARAM
+    : LIMIT PARAM
+    ;
+
+limitCommand_NAMED_PARAM
+    : LIMIT NAMED_PARAM
+    ;
+
+limitCommand_POSITIONAL_PARAM
+    : LIMIT POSITIONAL_PARAM
+    ;
```

Comment:
I was thinking about a similar question when I added these rules - where do we want to support parameters? There are many subcommands/clauses defined here:

`sourceCommand` - defines the objects to query against, perhaps it is less common to specify objects as parameter markers? 
`processingCommand` - I can imagine very common use cases to specify parameter markers in `whereCommand` and `evalCommand`, as users often code literal values in predicates.

It is less common that users specify parameter markers for a table/document name or a column/field name in my opinion(according to the SQL world). I'm thinking one option is to reduce the scope of this PR to support parameter markers in where and eval only. As we find more common use cases in other clauses, we can have a follow up/enhancement PR for them.

Reply:
`sourceCommand` - this one is in flux (see https://github.com/elastic/elasticsearch/pull/108395) and if the change will be merged it might be a candidate for a param usage (the source of the sourceCommand becomes now a regular string from the quoting point of view), but at the same time the name of the source (index name, pattern or CCS pattern) can be considered "equal" to a field name and field names are out of the scope of this feature.

In general, the params should be used in places where constants/literals can be used as well. I wouldn't say that a `processingCommand` or `evalCommand` is the right scope here. Imagine a function is using a param, wherever that function can be used, that's a scope boundary. For example:

```
    "query": "row x = \"Elastic\", y = ?1 | eval z = concat(x, ?1) | where starts_with(concat(x, ?1), x) | sort starts_with(concat(x, ?1), x)",
    "params": ["search"]
```

In the query above, the main actor is `concat` which can be used in different commands.
So, try to look at this from this angle, rather than a command-centric approach.
Indeed, `grok` and maybe `dissect` are separate commands that may not support this (and you need to dig into the "why") but for the rest this should be ok.

Reply:
IMO, the params allow for running "the same" query, over the same index (pattern), with same general characteristics (i.e. max row count). Along the lines of "I have a GUI showing a number of products, I can choose any of them to get some reporting. The product (name/id/etc) is a parameter."
In this view, changing the index name and count is unnecessary and I'd restrict the scope of the PR to it, until we learn we need more. I.e. extend current capabilities _only_.

Reply:
> IMO, the params allow for running "the same" query, over the same index (pattern), with same general characteristics (i.e. max row count). Along the lines of "I have a GUI showing a number of products, I can choose any of them to get some reporting. The product (name/id/etc) is a parameter." In this view, changing the index name and count is unnecessary and I'd restrict the scope of the PR to it, until we learn we need more. I.e. extend current capabilities _only_.

This is a common use case in SQL, there is dynamic statement with parameter markers on predicates and allow users to provide different values for each execution. One advantage of dynamic statement(SQL) with parameter markers is that it is prepared once, which means its access path is generated once at the first execution, and the subsequent executions have the option to reuse the same access path, it saves the time to optimize the query for each every execution, it is more efficient for OLTP queries to run this way, as they are usually short running queries. For analytical queries it might be a little different, as the execution time is much longer comparing to prepare time. 

We can provide more flexibilities to support parameter markers beyond predicates, when we see more requirements.